providers/fips/fipsprov.c

```diff
@@ -1238,3 +1246,163 @@ void OSSL_INDICATOR_get_callback(OSSL_LIB_CTX *libctx,
+
+        unset_key = true;
+
+        if (c_stcbfn != NULL && c_get_libctx != NULL)
+            c_stcbfn(c_get_libctx(FIPS_get_core_handle(libctx)), &cb, &cb_arg);
+
+        if ((ev = OSSL_SELF_TEST_new(cb, cb_arg)) == NULL)
+            goto done;
+
+        /* Mark test as in progress */
+        test->state = FIPS_DEFERRED_TEST_IN_PROGRESS;
+
+        /* execute test */
+        if (SELF_TEST_kats_single(ev, libctx, test->category, test->algorithm))
+            ret = FIPS_DEFERRED_TEST_PASSED;
+
+    done:
+        /* Mark test as pass or fail */
+        test->state = ret;
+
+        if (ev)
+            OSSL_SELF_TEST_free(ev);
+        if (unset_key)
+            CRYPTO_THREAD_set_local_ex(CRYPTO_THREAD_LOCAL_FIPS_DEFERRED_KEY,
+                                       libctx, NULL);
+        CRYPTO_THREAD_unlock(deferred_lock);
+    }
+    return ret;
+}
+
+static void deferred_test_error(int category)
+{
+    const char *category_name = "Unknown Category Test";
+
+    switch (category) {
+    case FIPS_DEFERRED_KAT_INTEGRITY:
+        category_name = OSSL_SELF_TEST_TYPE_KAT_INTEGRITY;
+        break;
+    case FIPS_DEFERRED_KAT_CIPHER:
+        category_name = OSSL_SELF_TEST_TYPE_KAT_CIPHER;
+        break;
+    case FIPS_DEFERRED_KAT_ASYM_CIPHER:
+        category_name = OSSL_SELF_TEST_TYPE_KAT_ASYM_CIPHER;
+        break;
+    case FIPS_DEFERRED_KAT_ASYM_KEYGEN:
+        category_name = OSSL_SELF_TEST_TYPE_KAT_ASYM_KEYGEN;
+        break;
+    case FIPS_DEFERRED_KAT_KEM:
+        category_name = OSSL_SELF_TEST_TYPE_KAT_KEM;
+        break;
+    case FIPS_DEFERRED_KAT_DIGEST:
+        category_name = OSSL_SELF_TEST_TYPE_KAT_DIGEST;
+        break;
+    case FIPS_DEFERRED_KAT_SIGNATURE:
+        category_name = OSSL_SELF_TEST_TYPE_KAT_SIGNATURE;
+        break;
+    case FIPS_DEFERRED_KAT_KDF:
+        category_name = OSSL_SELF_TEST_TYPE_KAT_KDF;
+        break;
+    case FIPS_DEFERRED_KAT_KA:
+        category_name = OSSL_SELF_TEST_TYPE_KAT_KA;
+        break;
+    }
+    ossl_set_error_state(category_name);
+}
+
+int FIPS_deferred_self_tests(OSSL_LIB_CTX *libctx, FIPS_DEFERRED_TEST tests[])
+{
+    int i;
+
+    /*
+     * NOTE: that the order in which we check the 'state' here is not important,
+     * if multiple threads are racing to check it the worst case scenario is
+     * that they will all try to run the tests. Proper locking for preventing
+     * concurrent tests runs and saving state from multiple threads is handled
+     * in FIPS_kat_deferred() so this race is of no real consequence.
+     */
+    for (i = 0; tests[i].algorithm != NULL; i++) {
+        if (tests[i].state != FIPS_DEFERRED_TEST_PASSED) {
+            int state;
+
+            /* any other threads that request a self test will lock and wait */
+            state = FIPS_kat_deferred(libctx, &tests[i]);
+            switch (state) {
+            case FIPS_DEFERRED_TEST_IN_PROGRESS:
+                /* self test in progress for this thread so we stop here. */
```

Comment:
I think this sums up the situation better:
```suggestion
                /* self tests in progress so we assume they pass and return success. */
```

However, I'm not sure this is correct behaviour.  Blocking until the self tests either pass or fail seems more reasonable.  Doing that would remove this state I suspect.

Reply:
A comment about locking is above the for loop, do you mean an additional comment to also explain that we let one thread go to complete the ongoing tests ?
I thought this comment would hint at that but I welcome clearer language.

Reply:
How about:
```
                /*
                 * The self tests are currently running.  If we are the thread that is running them
                 * then the algorithm is allowed so return okay.  Without this, deadlock will occur.
                 */
```

Reply:
we raced on it, let me know if the fixup I just push is good enough

Reply:
That works.